templates/frontend/auth/login.html.twig

```diff
@@ -6,10 +6,23 @@
         <div class="alert-error mb-4">{{ error.messageKey|trans(error.messageData, 'security') }}</div>
     {% endif %}
     <form action="{{ app.request.requestUri }}" method="post">
-        <div class="form-row">
-            <label for="username" class="labels">{{ 'username'|trans }}</label>
-            <input type="text" id="username" name="_username" value="{{ last_username }}" placeholder="{{ 'username'|trans }}">
-        </div>
+        {% if userProvider == "username" or userProvider is null %}
```

Comment:
A lot of duplicate html while only the label/placeholder are different, maybe extract the label? And I'd have the username/null case as a default at the end:
```twig
{% set identifierLabel %}
    {% if userIdentifier == 'email' %}
        {{ 'email'|trans }}
    {% elseif userIdentifier == 'both' %}
        {{ 'username_and_email'|trans }}
    {% else %}
        {{ 'username'|trans }}
    {% endif %}
{% endset %}
<div class="form-row">
    <label for="username" class="labels">{{ identifierLabel }}</label>
    <input type="text" id="username" name="_username" value="{{ last_username }}" placeholder="{{ identifierLabel }}">
</div>
```